workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/AndroidViewContainer.kt

```diff
@@ -0,0 +1,29 @@
+package com.squareup.workflow1.ui
+
+import android.content.Context
+import android.view.View
+import android.view.ViewGroup
+
+@OptIn(WorkflowUiExperimentalApi::class)
+class AndroidViewContainer<V : ViewRendering>(
```

Comment:
I always imagined this as an interface, but I think I understand why that doesn't work and this is better. Do I understand correctly:

The infra will use the vals on this type to create the initial view to show the rendering. On subsequent updates, they will use the usual mechanism to try to update the old view with the new rendering. If this fails, then they'll ask the viewFactory in this class to create a new view. This means that if a workflow returns an `AndroidViewContainer<String>` in one render pass, and an `AndroidViewContainer<Int>` in the next render pass, the view will be appropriately recreated with the `Int` view factory the second time around.

I think this _could_ be an interface if we had self types. Then we could do something like:
```kotlin
interface AndroidViewRendering {
  val viewFactory: ViewFactory<Self>
}
```

The way this would look with fake self-types is:
```kotlin
interface AndroidViewRendering<R : AndroidViewRendering<R>> {
  val viewFactory: ViewFactory<R>
}
```

So for example:
```kotlin
data class Rendering(
  val message: String,
  val onClick: () -> Unit
) : AndroidViewRendering<Rendering> {
  // This is kind of awkward, maybe it should just be a fun.
  override val viewFactory get() = LayoutRunner.bind(HelloGoodbyeLayoutBinding::inflate) { r, _ ->
    helloMessage.text = r.message
    helloMessage.setOnClickListener { r.onClick() }
  }
}
```

That seems less boilerplaty to me than having to create a function just to return the correct `AndroidViewContainer` type. WDYT?

Reply:
Ass written, the second AVC would never be shown. To fix that would have to make AVC implement Compatibile. I think that's another reason to avoid the wrapper and go instead with a base class or something.

Reply:
As I've written it, the second AVC would never be shown. To fix that would have to make AVC implement Compatibile. I think that's another reason to avoid the wrapper and go instead with a base class or something.

I hadn't thought about making a new interface, was trying to just make ViewFactory work, and I think it was the unrecognized need for the self typing that derailed me.

I'll try that and the base class, see how they feel.

Reply:
> I hadn't thought about making a new interface

Well, you already made `ViewRendering`, so what's one more? 😜 

Joking aside, I'm curious to see what you come up with. Renderings implementing an interface seems like the most straightforward API to me. I'm not sure what advantages you see from an abstract class over an interface.

Reply:
Yeah, sleeping on it (more or less) and looking again at your sketch, the interface seems like a clear win.

Reply:
Also, I don't see any reason for that `val viewFactory` to use `get()`.

Reply:
OMG SO GOOD

```kotlin
@OptIn(WorkflowUiExperimentalApi::class)
data class HelloView(
  val message: String,
  val onClick: () -> Unit
) : AndroidViewRendering<HelloView> {
  override val viewFactory: ViewFactory<HelloView> =
    LayoutRunner.bind(HelloGoodbyeLayoutBinding::inflate) { r, _ ->
      helloMessage.text = r.message
      helloMessage.setOnClickListener { r.onClick() }
    }
}
```

```kotlin
  override fun render(
    props: Unit,
    state: State,
    context: RenderContext
  ): HelloView {
    return HelloView(
      message = state.name,
      onClick = { context.actionSink.send(helloAction) }
    )
  }
```

Reply:
I wish I could delete the type declaration on the `val viewFactory` implementation, cant understand why it breaks inference. Oh well.